### PR TITLE
Metrics for the DB Store

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Replicate a version of what the server does to track DB Store metrics. This pull request includes discrete commits that refactor the existing `store` into `sqlstore` and `timerlayer` packages, as well as the containing `store` layer with the interface in question.

#### Ticket Link
None.

### CODE REVIEW

File: server/plugin.go [hunk 1]
```diff
@@ -165,12 +167,6 @@ func (p *Plugin) start(syncSince *time.Time) {
...
 
 	if enableMetrics != nil && *enableMetrics {
-		p.metricsService = metrics.NewMetrics(metrics.InstanceInfo{
-			InstallationID: os.Getenv("MM_CLOUD_INSTALLATION_ID"),
-		})
-
-		// run metrics server to expose data
-		p.runMetricsServer()
 		// run metrics updater recurring task
 		p.runMetricsUpdaterTask(p.store, updateMetricsTaskFrequency)
```
Comment (lieut-data) [inline server/plugin.go:171]: But we keep only starting the updater task on `start`.

File: server/plugin.go [hunk 2]
```diff
@@ -332,13 +339,14 @@ func (p *Plugin) OnActivate() error {
...
 
-		p.store = store.New(
+		store := sqlstore.New(
 			db,
 			p.apiClient.Store.DriverName(),
 			p.API,
 			func() []string { return strings.Split(p.configuration.EnabledTeams, ",") },
 			func() []byte { return []byte(p.configuration.EncryptionKey) },
 		)
+		p.store = timerlayer.New(store, p.metricsService)
```
Comment (lieut-data) [inline server/plugin.go:349]: Wrap the `sqlstore` with the `timerlayer`.

File: server/store/layer_generators/main.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (lieut-data) [inline server/store/layer_generators/main.go:1]: This is mostly just a simplified version of what's in the server.

File: server/store/layer_generators/main.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+	file.Close()
+	f, err := parser.ParseFile(fset, "", src, parser.AllErrors|parser.ParseComments)
+	if err != nil {
+		return nil, err
+	}
+
+	topLevelFunctionsToSkip := map[string]bool{
+		"BeginTx":    true,
+		"RollbackTx": true,
+		"CommitTx":   true,
```
Comment (lieut-data) [inline server/store/layer_generators/main.go:112]: We don't track these methods. (Ideally, these wouldn't be exposed outside `sqlstore` at all, but not touching that in this PR.)

File: server/store/storemodels/storemodels.go [hunk 5]
```diff
@@ -2,6 +2,12 @@ package storemodels
 
 import "time"
 
+type Stats struct {
+	ConnectedUsers int64
+	SyntheticUsers int64
+	LinkedChannels int64
+}
+
```
Comment (lieut-data) [inline server/store/storemodels/storemodels.go:10]: Moved this to the `storemodels` even though it's not persisted to simplify the import story.

**Review by jespino — APPROVED** [review]
LGTM. @manojmalik20 I know you were working on things around it, so just to keep you in the loop.

File: server/store/layer_generators/main.go [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+			return strings.Join(results, ", ")
+		},
+		"joinResultsForSignature": func(results []string) string {
+			if len(results) == 0 {
+				return ""
+			}
+			returns := []string{}
+			for _, result := range results {
+				switch result {
+				case "*PostReminderMetadata":
```
Comment (manojmalik20) [inline server/store/layer_generators/main.go:153]: When does this case occur?

File: server/store/layer_generators/timer_layer.go.tmpl [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	elapsed := float64(time.Since(start)) / float64(time.Second)
+        success := "false"
+        if {{$element.Results | errorToBoolean}} {
+                success = "true"
+        }
+        s.metrics.ObserveStoreMethodDuration("Store.{{$index}}", success, elapsed)
+	{{ with (genResultsVars $element.Results false ) -}}
+	return {{ . }}
+	{{- else -}}
+	}
```
Comment (manojmalik20) [inline server/store/layer_generators/timer_layer.go.tmpl:42]:
```suggestion
```
Can you verify if there's a need for this `else` condition or not?

**Review by manojmalik20 — APPROVED** [review]

File: server/store/layer_generators/main.go (hunk 6 above)
Comment (lieut-data) [reply server/store/layer_generators/main.go:153]: It doesn't! Good catch -- copy and paste error on my part :)

File: server/store/layer_generators/timer_layer.go.tmpl (hunk 7 above)
Comment (lieut-data) [reply server/store/layer_generators/timer_layer.go.tmpl:42]: I believe this handles the case when the store method returns nothing. I can't find a case where that happens today, but I think it makes sense for the layer generator to handle it seamlessly regardless.

Comment (manojmalik20) [reply server/store/layer_generators/timer_layer.go.tmpl:42]: When I was working on this template file for the client package, I got error here. If a function is not returning anything, then this `else` condition adds an extra `}` that gives error.

Comment (lieut-data) [reply server/store/layer_generators/timer_layer.go.tmpl:42]: Oh, I see! My apologies, I didn't follow the thread fully.

**Pull request merged**